### ISSUE TITLE
JSON payload with '%123' is blocked by rule -920271     msg:'Invalid character in request (non printable characters)',\

### ISSUE THREAD

**Issue by afekson** [coreruleset/coreruleset#2681]
### Description
In a POST request, when the JSON payload contains '%123', the request is blocked with 403 due to rule matching 920271 - Invalid character in request (non printable characters) 

Message :Invalid character in request (non printable characters) 
instanceId_s :appgw_0
SourceSystem :Azure
ruleSetType_s :OWASP CRS
ruleSetVersion_s: 3.2
ruleId_s :920271
ruleGroup_s: REQUEST-920-PROTOCOL-ENFORCEMENT
action_s: Matched
details_message_s
Validate Byte Range 0000000001100100000000000000000011111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111011111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111111 at ARGS.
details_data_s
ARGS:query=%123
details_file_s
REQUEST-920-PROTOCOL-ENFORCEMENT.conf

### Audit Logs / Triggered Rule Numbers

ruleId_s :920271
ruleGroup_s: REQUEST-920-PROTOCOL-ENFORCEMENT

### Your Environment

engine_s :Azwaf 
SourceSystem :Azure
### Confirmation

[x] I have removed any personal data (email addresses, IP addresses,
    passwords, domain names) from any logs posted.

**Comment by airween** [comment]
Hi @targonel13,

thanks for feedback! Unfortunately this is a false positive hit. We will investigate this issue soon, but if you need a quick solution, you can create an exclusion rule, for eg.:

```
SecRule REQUEST_URI "@beginwsWith /your/uri" \
    "id:990001,\
    phase:1,\
    t:none:\
    pass,\
    nolog,\
    ctl:ruleRemoveByID=920271"
```

**Comment by airween** [comment]
Hi @targonel13,

Can you share with us a `curl` request, which we can reproduce your issue?

I've made some researches, and found that this request didn't blocked by rule 920271:

```
curl -v -d '{"query":"%25123"}' -H 'Content-Type: application/json; charset=utf-8' http://localhost
```
In debug.log, I can see the decoded form of the request:
```
[4] Executing operator "validateByteRange" with param "9,10,13,32-126,128-255" against ARGS:query.
[9] Target value: "%123"
[4] Operator completed in 0 usec.
[4] Rule returned 0.
```

So I'm afraid your request uses wrong encoding. Please help me to clarify it.

**Comment by targonel13** [comment]
Hi,
This is the curl


curl -X 'POST' \
  'https://<uriprefix>/query?api-version=2022-06-01-draft' \
  -H 'accept: text/plain' \
  -H 'Authorization: Bearer xxxxxxxxx
  -H 'Content-Type: application/json' \
  -d '{
  "query": "%123"
}'

**Comment by airween** [comment]
Hi @targonel13,

```
curl -X 'POST' \
 'https://<uriprefix>/query?api-version=2022-06-01-draft' \
 -H 'accept: text/plain' \
 -H 'Authorization: Bearer xxxxxxxxx -H 'Content-Type: application/json' \
 -d '{ "query": "%123" }'
```
thanks - as I wrote above, in this case you should encode your JSON data. The correct form is this:

```
 -d '{ "query": "%25123" }'
```
As you can see, you have to replace the `%` by the `%25` (`%` char's code is 37 in decimal, which is 25 in hex). Most programming language able to do that, eg. in Python, you can check:
```python
>>> import urllib.parse
>>> urllib.parse.quote_plus("%123")
'%25123'
```

**Comment by targonel13** [comment]
Hi @airween , 
Let me explain a bit more about the situation.
I am creating  a public API to allow the user to search files.
The user is sending me the query using the Query API.
I am storing files for the client and want to allow the client to search for them.
Suppose i have a stored file named '%123.txt' in my system-
for example the query could be  "query": "FileName == '%123.txt'".


My service is hosted in Azure and i do not have any place to replace any notes, and i don't want to since i want to take the exact string and compare it with my sorted files names for example.

The call is being blocked in the WAF so i have no access the the content since it does not reach my backend and blocked by this rule.

**Comment by airween** [comment]
Hi @targonel13,

thanks for the summary.

> I am creating a public API to allow the user to search files. The user is sending me the query using the Query API.
> ...
> Suppose i have a stored file named '%123.txt' in my system- for example the query could be "query": "FileName == '%123.txt'".

I see, but above I explained that the query of your user, the request itself is not valid.

You or your user have to encode the query - that's the solution.

> i want to take the exact string and compare it with my sorted files names for example.

Agree, but please note that if you want to send the `%` character in HTTP, you have to encode it, because `%` has a special meaning. For more info, please read this [article](https://en.wikipedia.org/wiki/Percent-encoding).

**Comment by targonel13** [comment]
Thanks for the response.
Let me see if i understand it correctly,
In the body of the http POST call, it is forbitten to have '%111' in the body of the request, and if the client insists of having files with names containing '%123' or similar he should replace the % with %25

Also, i understand the this called being blocked is by design and no intention to change it, right?

**Comment by airween** [comment]
> In the body of the http POST call, it is forbitten to have '%111' in the body of the request, and if the client insists of having files with names containing '%123' or similar he should replace the % with %25

Yes. You should consider this, if the server receives an argument which contains a `%`, then it looks up the next two characters, which must be numbers. The server will decode this sequence: cuts the `%`, and generates a hexa number from the next two numbers. In your case, the `%12` describes the character ASCII 18 (hex 12 = dec 18), which is a non-printable character - therefore that rule does not allow it.

> Also, i understand the this called being blocked is by design and no intention to change it, right?

Yes, exactly.

**Comment by aelij** [comment]
@airween Can you explain the rationale for requiring this encoding in JSON request bodies? Unlike URIs, normally JSON payloads are not URI-encoded (or decoded). It does make sense if the body's content was `application/x-www-form-urlencoded`.

**Comment by airween** [comment]
Hi @aelij,

I've made a quick research, and may be you're right. And this means that rule gives false positive in this case.

We have to discuss it - we will be in touch soon.

Thanks again.

**Comment by azurit** [comment]
@airween Any news here? Have this already been discussed? Thanks for the update!

**Comment by dune73** [comment]
Let me add my two cents.

920271 is a paranoia level rule. That means false positive can happen.

920271 also dies URLDecoding. `%123` is thus `%12` + `3`. Now %12 is Ascii 18, device control 2, and thus an invisible escape sequence and exactly the character class that 920271 is meaning to forbid.

So under the line I do not see a chance we are fixing this. You need to treat this as a rule exclusion.

I'll leave this open for now, but unless further notice, we'll close this issue.

**Comment by azurit** [comment]
More than 2 years have passed - I suggest to close this.
